src/python/pants/engine/target.py

```diff
@@ -1352,13 +1353,41 @@ class NoApplicableTargetsBehavior(Enum):
     error = "error"
 
 
+def parse_shard_spec(shard_spec: str) -> Tuple[int, int]:
+    def invalid():
+        return ValueError(
+            f'Invalid shard specification {shard_spec}. Use a string of the form "k/N" where '
+            "k and N are integers, and 0 <= k < N ."
```

Comment:
It will be confusing where this error message is coming from. How about having callers pass through the corresponding option name? 

So this would read:

> Invalid shard specification `{shard_spec}` from the option `[test].shard_spec`.

Reply:
If you don't want to wire that up to this function, could have it return `tuple[int, int] | None`, then callers raise the error.

Reply:
Easier to move the function call out of the ctor and into the level up. Done.